### PR TITLE
Changes to run tests on beta builds

### PR DESCRIPTION
**Pull request by isabelrios**
This PR shows the modifications needed to run tests on beta builds on Firebase

### CODE REVIEW

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 1]
```diff
@@ -50,12 +50,6 @@ class BrowserRobot {
     }
 
     fun verifyUrl(url: String) {
-        val mDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
-        mDevice.waitNotNull(
-            Until.findObject(By.res("org.mozilla.fenix.debug:id/mozac_browser_toolbar_url_view")),
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:55]: We will need to modify all these calls so that tests also work on beta. This only works on debug. But if we use just the id like `onView(withId(R.id.mozac_browser_toolbar_url_view))` it will work on beta and nightly

**Comment by isabelrios** [comment]
@AaronMT @rpappalax just FYi so that you are aware of the changes I have been doing. Not all the changes related to the tests are applied here but some examples of changes needed to have all the tests working on Beta. Hope this helps

File: app/proguard-rules.pro [hunk 2]
```diff
@@ -1,5 +1,5 @@
 -dontobfuscate
-
+-dontoptimize
```
Comment (pocmo) [inline app/proguard-rules.pro:2]: Does it work for you too if you don't add this, but have `minifyEnabled` (and maybe `shrinkResources`) set to `false`? (Just curious since that may be easier to control in automation)
https://github.com/mozilla-mobile/fenix/blob/master/app/build.gradle#L49-L50

Comment (isabelrios) [reply app/proguard-rules.pro:2]: Thanks @pocmo!! looks like that may work... setting both to false no need to add `-dontoptimize` to get tests running

Comment (pocmo) [reply app/proguard-rules.pro]: Thank you for confirming! I created a PR to make this configurable via a build flag: https://github.com/mozilla-mobile/fenix/pull/16787

File: taskcluster/ci/ui-test/kind.yml [hunk 3]
```diff
@@ -64,3 +64,31 @@ jobs:
...
+        attributes:
+            build-type: debug
+            code-review: true
+            retrigger: true
+        dependencies:
+            signing: signing-beta-firebase
+            signing-android-test: signing-android-test-beta
+        run:
+            commands:
+                - [wget, {artifact-reference: '<signing-beta-firebase/public/build/x86/target.apk>'}, '-O', app.apk]
```
Comment (isabelrios) [inline taskcluster/ci/ui-test/kind.yml:77]: @JohanLorenzo it is only sigining for debug, like the -debug is appended...but for beta we need to specify, would that be the way? I used the naming after the task name

Comment (isabelrios) [reply taskcluster/ci/ui-test/kind.yml:77]: Also, are these wget commands overwritten too? If not I should take them out fo the defaults above...

Comment (JohanLorenzo) [reply taskcluster/ci/ui-test/kind.yml:77]: You can keep `<signing/public/build/x86/target.apk>` (and `<signing-android-test/public/build/noarch/target.apk>`). Taskgraph will know you refer to the beta tasks thanks to the mapping you provide in `dependencies` 

> Also, are these wget commands overwritten too? If not I should take them out fo the defaults above...

That's an excellent question! It's a behavior of taskgraph that I don't like. Taskgraph doesn't overwrite lists like `commands`. It concatenates them, instead:
 * https://hg.mozilla.org/ci/taskgraph/file/2b2622598df02bde211d8cedb334b7b22fb883a4/src/taskgraph/loader/transform.py#l47
 * https://hg.mozilla.org/ci/taskgraph/file/2b2622598df02bde211d8cedb334b7b22fb883a4/src/taskgraph/util/templates.py#l33

We need to take out `commands` from `job-defaults` :/

Comment (isabelrios) [reply taskcluster/ci/ui-test/kind.yml:77]: Okay! thanks I will do that!

File: taskcluster/ci/ui-test/kind.yml [hunk 4]
```diff
@@ -64,3 +64,16 @@ jobs:
                 - [automation/taskcluster/androidTest/ui-test.sh, x86-screenshots-tests, app.apk, android-test.apk, '-1']
         treeherder:
             symbol: debug(screenshots-x86)
+    x86-beta:
+        attributes:
+            build-type: beta
```
Comment (JohanLorenzo) [inline taskcluster/ci/ui-test/kind.yml:69]: Nit:

```suggestion
            build-type: beta-firebase
```

Let's match the build type we defined here:
https://github.com/mozilla-mobile/fenix/blob/7a08175568fe2a7e426b8542554a3a295aab5189/taskcluster/ci/build/kind.yml#L57

File: taskcluster/ci/ui-test/kind.yml [hunk 5]
```diff
@@ -64,3 +64,16 @@ jobs:
...
+        description: Test Fenix
+        run-on-tasks-for: [github-push]
+        dependencies:
+            signing: signing-beta-firebase
+            signing-android-test: signing-android-test-beta
+        run:
+            commands:
+                - [automation/taskcluster/androidTest/ui-test.sh, x86-beta-tests, app.apk, android-test.apk, '50']
+        treeherder:
+            symbol: debug(ui-test-x86-beta)
```
Comment (JohanLorenzo) [inline taskcluster/ci/ui-test/kind.yml:79]: Nit:

```suggestion
            symbol: beta(ui-test-x86-beta)
```

This way, it will appear alonside the other beta jobs: https://treeherder.mozilla.org/jobs?repo=fenix&revision=d4ef63f8778f4458565452661bfc6e22766bcd5c&selectedTaskRun=F6aaOEelTrGRcttgoKTRXw.0

File: try_task_config.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (JohanLorenzo) [inline try_task_config.json:1]: TODO before landing: remove that file 🙂

**Review by JohanLorenzo — APPROVED** [review]
The taskcluster bits looks great to me, modulo a few nits! 

Thanks a lot for picking this up super quickly! 💯

File: automation/taskcluster/androidTest/flank-x86-beta.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  auto-google-login: true
+  use-orchestrator: true
+  environment-variables:
+    clearPackageData: true
+  directories-to-pull:
+    - /sdcard/screenshots
+  performance-metrics: true
+
+  test-targets:
+   - class org.mozilla.fenix.ui.SmokeTest
```
Comment (isabelrios) [inline automation/taskcluster/androidTest/flank-x86-beta.yml:41]: @AaronMT @rpappalax would we start with this or just one or two tests? wdyt?

File: app/build.gradle [hunk 8]
```diff
@@ -550,7 +550,7 @@ dependencies {
     androidTestImplementation Deps.androidx_junit
     androidTestImplementation Deps.androidx_work_testing
     androidTestImplementation Deps.mockwebserver
-
+    implementation("com.squareup.okhttp3:okhttp:4.9.0")
```
Comment (pocmo) [inline app/build.gradle:553]: Drive-by: Is this really needed? 🤔 - Since we are not using (and do not want to use) this in the app, I wonder if `testImplementation` would be enough. But still curious if needed at all?

Comment (isabelrios) [reply app/build.gradle]: I needed that while I was debugging to get the tests running, but you are right, looks like we can remove this...

**Review by AaronMT — APPROVED** [review]

**Comment by isabelrios** [comment]
@JohanLorenzo @pocmo sorry guys, asking for your review again due to all the recent changes since you last reviewed it. Thank you in advance!!

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt (hunk 1 above)
Comment (rpappalax) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:60]: @isabelrios how will we handle this so that we don't need 2 versions of each test?  If I recall correctly didn't @AaronMT or @sv-ohorvath say there was a way to parametrize the package name (if it's needed)?

Comment (rpappalax) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:60]: nvrmd, i see that you're one step ahead of me :)

File: automation/taskcluster/androidTest/flank-x86-beta.yml (hunk 7 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/flank-x86-beta.yml:41]: @isabelrios either way.  Let's start with 1-3 tests that are reliable

**Comment by rpappalax** [comment]
+lgtm 👍🏼

File: automation/taskcluster/androidTest/flank-x86-beta.yml (hunk 7 above)
Comment (isabelrios) [reply automation/taskcluster/androidTest/flank-x86-beta.yml:41]: I think I will pick the `NavigationToolbarTest#visitURLTest` to start with and then we can add the smoketest or more tests as we decide

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt (hunk 1 above)
Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:60]: yeah, that was already implemented and working nicely :) with that we will be able to run tests using any package, debug, beta or nightly

**Comment by isabelrios** [comment]
As a summary, with this PR we will enable running tests on beta builds only on releases branches. 
- ui-test-x86-debug on master running all tests. It will not run on releases branches.
- ui-test-x86-beta will not run on master but it will run on releases branches executing only one test for now, then we will enable the smoketest or other stable tests.

File: taskcluster/ci/ui-test/kind.yml [hunk 9]
```diff
@@ -49,7 +49,7 @@ jobs:
     x86-debug:
         description: Test Fenix
         run-on-tasks-for: [github-pull-request, github-push]
-        run-on-git-branches: [master]
+        run-on-git-branches: ["^((?!releases_)(?!releases/).)*$"]
```
Comment (JohanLorenzo) [inline taskcluster/ci/ui-test/kind.yml:52]: Great idea on filtering out the release branches 👍 I hadn't thought of it 🙂

That made me realize we can simplify the regexes below and this one too. I think that one is equivalent: 

```suggestion
        run-on-git-branches: ["^((?!releases[_/]).+)$"]
```

File: taskcluster/ci/ui-test/kind.yml [hunk 10]
```diff
@@ -66,3 +66,17 @@ jobs:
                 - [automation/taskcluster/androidTest/ui-test.sh, x86-screenshots-tests, app.apk, android-test.apk, '-1']
         treeherder:
             symbol: debug(screenshots-x86)
+    x86-beta:
+        attributes:
+            build-type: beta-firebase
+        description: Test Fenix
+        run-on-tasks-for: [github-push]
+        run-on-git-branches: ["^releases/.+$", "^releases_.+$"]
```
Comment (JohanLorenzo) [inline taskcluster/ci/ui-test/kind.yml:74]: I'm sorry, I didn't provide the best regex. This one is equivalent and looks simpler to me.

```suggestion
        run-on-git-branches: ["^releases[/_].+$"]
```

**Review by JohanLorenzo — APPROVED** [review]
The taskcluster parts look to me. Great call on filtering out releases branches! I think we can provide what I believe are simpler regexes, but I guess it's a matter of personal preference. Feel free to not include my suggestions if you don't think my regexes look simpler to you.

File: app/build.gradle [hunk 11]
```diff
@@ -550,7 +550,6 @@ dependencies {
     androidTestImplementation Deps.androidx_junit
     androidTestImplementation Deps.androidx_work_testing
     androidTestImplementation Deps.mockwebserver
-
```
Comment (pocmo) [inline app/build.gradle:553]: nit: there's no change except the new line removal here anymore :)

File: app/proguard-rules.pro [hunk 12]
```diff
@@ -119,4 +119,5 @@
 
 # Keep Android Lifecycle methods
 # https://bugzilla.mozilla.org/show_bug.cgi?id=1596302
--keep class androidx.lifecycle.** { *; }
\ No newline at end of file
+-keep class androidx.lifecycle.** { *; }
+-keep class okhttp3.** { *; }
```
Comment (pocmo) [inline app/proguard-rules.pro:123]: nit: since we didn't need to add okhttp as a dependency, I assume we will not need the keep rule either?

**Review by pocmo — APPROVED** [review]

**Comment by st3fan** [comment]
@isabelrios just a quick note that we also need to update the rule in `.mergify.yml` so that it blocks auto-landing if the UI tests fail. That can be added to this PR or separately. We would need that in the `Release Automation (New)` rule.

**Comment by st3fan** [comment]
Here is the Mergify update to include the `ui-test-x86-beta` check as a condition for auto-landing the automated PRs.

https://github.com/mozilla-mobile/fenix/pull/17126

**Comment by isabelrios** [comment]
As discussed on Slack, let's wait to land this PR until Monday just in case there are build issues

**Comment by isabelrios** [comment]
@Mergifyio backport releases_v85.0.0
